File: cpp/cmd/cat.cpp
// SPDX-License-Identifier: MIT
#include <koru/braam.hpp>

namespace {

constexpr Str USAGE =
    "Usage:\n"
    "    cat [<file>...]\n";

} // namespace

// Chunks, not lines: cat is byte-exact, so a last line without a newline stays
// that way. Named files are read end to end as one stream.
Task<i32> proc_main(Args args)
{
    if (help_asked(args))
        co_return co_await usage_asked(USAGE);

    Input files(args.tail(), SYS_STDIN, "cat");
    File in(files);
    File &out = File::stdout();

    // A span at a time. On the console the buffering stays a line.
    if (out.reserve(SYS_READ_MAX).is_err())
        co_return 1;

    char buf[256];
    for (;;) {
        Result<usize> n = co_await in.read(buf);
        if (n.is_err())
            break;
        if ((co_await out.write(Str(buf, n.value()))).is_err())
            break;
    }

    if ((co_await out.flush()).is_err())
        co_return 1;
    if (in.failed())
        co_return in.err() == Error::Cancelled ? 130 : 1;
    co_return 0;
}

File: rust/runtime/examples/cat.rs
// SPDX-License-Identifier: MIT

//! Braam's `src/cmd/cat.cpp`, in Rust. Chunks, not lines: `cat` is byte-exact,
//! so a last line without a newline stays that way.

use koru::{Args, File, Input, Kind, help_asked, usage_asked};

const USAGE: &str = "Usage:
    cat [<file>...]
";

#[koru::main]
async fn main(args: Args) -> i32 {
    if help_asked(&args) {
        return usage_asked(USAGE).await;
    }

    let mut input = File::over(Input::new(args.tail(), koru::stdin(), "cat"));
    let out = File::stdout();

    let mut buf = [0u8; 256];
    loop {
        let n = match input.read(&mut buf).await {
            Ok(n) => n,
            Err(_) => break, // the error sticks on the stream and is read below
        };
        if out.write_bytes(&buf[..n]).await.is_err() {
            break;
        }
    }

    if out.flush().await.is_err() {
        return 1;
    }
    if input.failed() {
        return match input.err() {
            Some(e) if e.is(Kind::Cancelled) => 130,
            _ => 1,
        };
    }
    0
}
